#14 interraction with tg bot api

klimenkoMA opened the pull request:
Close #14 
Created a convenient class for sending and receiving a response from the telegram API, and tested this class

klimenkoMA committed:
#14
- created HighLevelSendRequestService class

klimenkoMA committed:
#14 changes in HighLevelSendRequestService class
- created 2 overloading sendMessage() methods

klimenkoMA committed:
#14 changes in HighLevelSendRequestService class
- created 2 overloading getChat() methods

klimenkoMA committed:
Merge branch '#13-create-send-requests-service' into #14-InterractionWithTGBotAPI

klimenkoMA committed:
Merge branch '#13-create-send-requests-service' into #14-InterractionWithTGBotAPI

klimenkoMA committed:
Merge branch '#13-create-send-requests-service' into #14-InterractionWithTGBotAPI

klimenkoMA committed:
Merge branch '#13-create-send-requests-service' into #14-InterractionWithTGBotAPI

klimenkoMA committed:
#14 changes in HighLevelSendRequestService class
- created getMe() method
- created final String GET_ME
- created final Logger LOGGER

klimenkoMA committed:
#14 changes in HighLevelSendRequestService class
- created constructor with @Autowired SendRequestService field
- created HighLevelSendRequestServiceTest class
- created a method methodGetMeShouldReturnedJSONAnswer() that passed the test successfully

klimenkoMA committed:
Merge branch 'dev' into #14-InterractionWithTGBotAPI

klimenkoMA committed:
#14 changes in HighLevelSendRequestService class
- created enum TGApiRequest
- deleted final String GET_ME

klimenkoMA committed:
#14 changes in HighLevelSendRequestService class
- realized method getChat() with Integer argument chatId
- realized method getChat() with String argument chatId

in HighLevelSendRequestServiceTest class
- created integerMethodGetChatShouldReturnedJSONAnswer() test
- created stringMethodGetChatShouldReturnedJSONAnswer()
- all tests passed successfully

klimenkoMA committed:
#14 changes in HighLevelSendRequestService class
- renamed JSONObject argument keyboard in sendMessage()
- made some minor edits to make the code readable

klimenkoMA committed:
Merge branch 'dev' into #14-InterractionWithTGBotAPI

klimenkoMA committed:
#14 changes in HighLevelSendRequestService class
- realised sendMessage() without keyboard
- created CHAT_ID, TEXT in TGApiRequest enum

 in HighLevelSendRequestServiceTest class
 - created sendMessageWithOutKeyboardShouldReturnedJSONAnswer test, passed successfully

klimenkoMA committed:
#14 changes in HighLevelSendRequestService class
- rebuild sendMessage()

  Created OutgoingMessage class

 - rebuild sendMessageWithOutKeyBoardShouldReturnJSONAnswer() in HighLevelSendRequestServiceTest class

klimenkoMA committed:
Merge branch 'dev' into #14-InterractionWithTGBotAPI

# Conflicts:
#	src/main/java/com/example/telegrambotinsurance/service/SendRequestsService.java

klimenkoMA committed:
#14 changes in HighLevelSendRequestServiceTest class
- removed net.minidev.json, import org.json library
from HighLevelSendRequestService class

klimenkoMA committed:
#14 changes
- created a comments in HighLevelSendRequestService class, HighLevelSendRequestServiceTest class, OutgoingMessage class

klimenkoMA committed:
#14 changes
- wrote a "TODO" in HighLevelSendRequestService class

spasynkov commented on src/main/java/com/example/telegrambotinsurance/service/HighLevelSendRequestService.java:
```diff
@@ -0,0 +1,135 @@
+package com.example.telegrambotinsurance.service;
+
+
+//TODO implement the sendMessage method with the keyboard!!!
```
это нужно?
если нужно - то перенеси куда-то внутрь класса. вряд ли кто-то будет такую важную штуку в импортах искать

klimenkoMA replied:
Принял, перенесу

klimenkoMA replied:
Перенес

spasynkov commented on src/main/java/com/example/telegrambotinsurance/service/HighLevelSendRequestService.java:
```diff
@@ -0,0 +1,135 @@
+
+
+@Service
+public class HighLevelSendRequestService {
```
я предлагаю подумать над названием
может лучше `TelegramApi`?
тогда как-то так будет
```
TelegramApi api = new TelegramApi();
...
api.sendMessage(...);
```

ну в общем, как-то дать понять, что тут основные методы реализованы из апи

klimenkoMA replied:
Хорошо, переделаю

klimenkoMA replied:
Название переделал

spasynkov commented on src/main/java/com/example/telegrambotinsurance/service/HighLevelSendRequestService.java:
```diff
@@ -0,0 +1,135 @@
+@Service
+public class HighLevelSendRequestService {
+
+	private final Logger LOGGER = LoggerFactory.getLogger(HighLevelSendRequestService.class);
```
почему не статический? у каждого объекта будет свой логгер? :)

klimenkoMA replied:
Исправлю

klimenkoMA replied:
Добавил статический модификатор

spasynkov commented on src/main/java/com/example/telegrambotinsurance/service/HighLevelSendRequestService.java:
```diff
@@ -0,0 +1,135 @@
+	 * @param message message text
+	 * @return response from the API telegram
+	 */
+	public JSONObject sendMessage(AbstractBot bot, String token, OutgoingMessage message) {
```
а зачем тебе отдельно токен, если токен всегда можно достать из бота?

klimenkoMA replied:
В задании ты писал, что нужен бот и токен, причем в каждом методе класса

spasynkov replied:
аа, ну я там через слеш написал :) типа, или :)
ну просто не знал нужен ли будет бот для дальнейших каких-то действий, или достаточно будет просто токена...

в общем, давай оставим только бота. а токен уже из объекта бота доставать когда надо

klimenkoMA replied:
Хорошо, переделаю

klimenkoMA replied:
Реализовал работу всех методов без параметра String token

spasynkov commented on src/main/java/com/example/telegrambotinsurance/service/HighLevelSendRequestService.java:
```diff
@@ -0,0 +1,135 @@
+
+		JSONObject jsonObject = sendRequestsService.sendGet(token,
+				TGApiRequest.SEND_MESSAGE.getRequest()
+						+ TGApiRequest.CHAT_ID.getRequest() + message.getStringChatId() +
+						TGApiRequest.TEXT.getRequest() + message.getText());
```
нет. вот так мне не нравится.
одно дело вынести в енумы названия методов, это ок.
но вот выносить туда вообще все строки - это совсем не ок

если тебе нужно как-то "собирать" строку из названия метода и параметров - то надо делать это удобно, прозрачно.
даже так - уже лучше:
```
new StringBuilder()
.append(TGApiRequest.SEND_MESSAGE.text)
.append("?")
.append("chat_id=").append(message.getStringChatId())
.append("&")
.append("text=").append(message.getText())
.toString()
```
оно значительно читабельней, чем сейчас сделано у тебя. в таком виде четко видно как именно собирается строка с параметрами, из чего. намного проще потом будет что-то добавить/удалить.

ты сделал тоже прикольно, но это реально оверинжиниринг уже :) 
сложнее != лучше

spasynkov replied:
можешь сделать отдельным классом, где можно будет просто вызывать удобные методы, типа `addParam(key, value)`, который будет добавлять параметры эти, ставить разделительные знаки сам, итд

а еще лучше посмотреть как другие собирает строку с паарметрами. я уверен, что в джаве для такого уже есть готовые решения, и не надо будет свои велосипеды выдумывать

klimenkoMA replied:
Хорошо, переделаю

spasynkov commented on src/main/java/com/example/telegrambotinsurance/modelbot/OutgoingMessage.java:
```diff
@@ -0,0 +1,22 @@
+
+public class OutgoingMessage {
+
+	private String stringChatId;
+	private long longChatId;
```
а зачем это?
мы же говорили, что пусть будет всегда строка. тогда не будет конфликта если там будут цифры. цифры в строковом представлении тоже ок

klimenkoMA replied:
Я не знал, как в конструкторе сделать так, чтобы цифру переводить в строку. Но вроде придумал, попробую реализовать

spasynkov replied:
не, так с цифрами вообще не работать, в принципе :)
везде работать со строкой :) просто не принимать цифры в принципе. потому что во входящем сообщении там тоже вроде как в строковом виде это все будет (даже если там будут цифры - то это будет строка "12345")

spasynkov replied:
соответственно, и не придется делить на чат айди из цифер, и чат айди из строки. соответственно отпадает необходимость в дублировании кода.

если вот прям очень хочется чтоб была возможность передавать в твои методы цифры - то да, сделать конструктор, который будет принимать цифры, а внутри него переводить в строку и сохранять как строку уже :)

klimenkoMA replied:
Хорошо, понял, сделаю только строку

spasynkov commented on src/main/java/com/example/telegrambotinsurance/service/HighLevelSendRequestService.java:
```diff
@@ -0,0 +1,135 @@
+				TGApiRequest.SEND_MESSAGE.getRequest()
+						+ TGApiRequest.CHAT_ID.getRequest() + message.getStringChatId() +
+						TGApiRequest.TEXT.getRequest() + message.getText());
+		LOGGER.debug(TGApiRequest.SEND_MESSAGE.getRequest());
```
это точно надо? ведь у нас и так выводятся логи когда мы отправляем сообщение какое-то. с урлом полным

klimenkoMA replied:
Уберу, сделал так, как в низкоуровневом классе

spasynkov replied:
там вроде убрали тот лог. как-раз потому, что сама библиотека по отправке запросов - она логи пишет :) это я когда видео про дебаг записывал - то заметил этот момент

spasynkov commented on src/main/java/com/example/telegrambotinsurance/service/HighLevelSendRequestService.java:
```diff
@@ -0,0 +1,135 @@
+	 */
+	public JSONObject sendMessage(AbstractBot bot, String token, OutgoingMessage message) {
+
+		JSONObject jsonObject = sendRequestsService.sendGet(token,
```
а почему `sendGet` вообще, если в документации написано, что надо отправлять пост запрос?
то, что оно работает через гет - еще не значит, что так надо делать.

или это чтобы "палить" всю переписку всем желающим, кто захочет проснифить трафик?
телеграм даже роскомнадзору ключи шифрования не дает чтобы не палить переписку пользователей, а ты тут собираешься прям в урле текст сообщения отправлять? :)

klimenkoMA replied:
Нехорошо получилось, исправлю :)

spasynkov commented on src/main/java/com/example/telegrambotinsurance/service/HighLevelSendRequestService.java:
```diff
@@ -0,0 +1,135 @@
+	 */
+	public JSONObject sendMessage(AbstractBot bot, String token, OutgoingMessage message,
+	                              JSONObject keyboard) {
+		return new JSONObject();
```
это типа еще не сделано?
в таких случаях лучше кидать исключение. типа `new RuntimeException("Not implemented!")`
кроме того, у тебя этот метод не будет отличаться практически ничем от предыдущего метода, кроме как что в этом еще клавиатуру "добавлять" в джейсон надо будет.

ну так ведь тогда придется либо дублировать код (что плохо), либо уже сразу сейчас придумать как сделать так, чтобы код был в одном месте.

я предлагаю в том методе без клавиатуры вызвать этот, но передать в качестве клавиатуры - null
а здесь уже написать весь тот код из первого метода, и сделать иф, типа, если клавиатура не нулл - тогда - и вот тут уже оставить тудушку или кидать not implemented

klimenkoMA replied:
Принял, переделаю

spasynkov commented on src/main/java/com/example/telegrambotinsurance/service/HighLevelSendRequestService.java:
```diff
@@ -0,0 +1,135 @@
+	 */
+	public JSONObject getChat(AbstractBot bot, String token, String chatId) {
+		JSONObject jsonObject = sendRequestsService.sendGet(token,
+				TGApiRequest.GET_CHAT.getRequest() + TGApiRequest.CHAT_ID.getRequest()
+						+ chatId);
+		LOGGER.debug(TGApiRequest.GET_CHAT.getRequest());
```
все аналогично замечаниям для `sendMessage()`

klimenkoMA replied:
Хорошо, переделаю

spasynkov commented on src/main/java/com/example/telegrambotinsurance/service/HighLevelSendRequestService.java:
```diff
@@ -0,0 +1,135 @@
+	 * @param chatId long type chat ID
+	 * @return response from the API telegram
+	 */
+	public JSONObject getChat(AbstractBot bot, String token, long chatId) {
+		JSONObject jsonObject = sendRequestsService.sendGet(token,
+				TGApiRequest.GET_CHAT.getRequest() + TGApiRequest.CHAT_ID.getRequest()
+						+ chatId);
+		LOGGER.debug(TGApiRequest.GET_CHAT.getRequest());
+		return jsonObject;
+	}
```
не думаю, что этот метод нужен, если будет только одно строкове поле с айдишником чата

klimenkoMA replied:
Хорошо, уберу

spasynkov commented on src/main/java/com/example/telegrambotinsurance/service/HighLevelSendRequestService.java:
```diff
@@ -0,0 +1,135 @@
+	 * @param token a token of the bot
+	 * @return response from the API telegram
+	 */
+	public JSONObject getMe(AbstractBot bot, String token) {
+		JSONObject jsonObject = sendRequestsService.sendGet(token,
+				TGApiRequest.GET_ME.getRequest());
+		LOGGER.debug(TGApiRequest.GET_ME.getRequest());
+
+		return jsonObject;
+	}
```
аналогично замечаниям по `sendMessage`

klimenkoMA replied:
Принял, исправлю

spasynkov commented on src/main/java/com/example/telegrambotinsurance/service/HighLevelSendRequestService.java:
```diff
@@ -0,0 +1,135 @@
+	 * class for generality of composing the http request string
+	 */
+	private enum TGApiRequest {
+		GET_ME("getMe"), SEND_MESSAGE("sendMessage"), GET_CHAT("getChat"),
+		CHAT_ID("?chat_id="), TEXT("&text=");
```
если енумы простые, без параметров - то можно и в строчку их писать
но вообще если енумы принимают параметры - то каждую константу с новой строки лучше писать

как я уже писал выше `CHAT_ID` и `TEXT` здесь не нужны

klimenkoMA replied:
Хорошо, переделаю

spasynkov commented on src/main/java/com/example/telegrambotinsurance/service/HighLevelSendRequestService.java:
```diff
@@ -0,0 +1,135 @@
+		/**
+		 * this field contains the value of the enum used
+		 */
+		private final String request;
```
давай просто `text`
и короче, и понятнее, чем какой-то "запрос" (куда? зачем?)

klimenkoMA replied:
Окей, исправлю

spasynkov commented on src/main/java/com/example/telegrambotinsurance/service/HighLevelSendRequestService.java:
```diff
@@ -0,0 +1,135 @@
+		 * this method returns value of the enum used
+		 * @return value of  the enum used
+		 */
+		public String getRequest() {
+			return request;
+		}
```
а этот метод вообще не нужен, если енум у тебя внутренний класс. внешний класс "видит" приватные переменные внутреннего, и наоборот

klimenkoMA replied:
Хорошо, уберу

spasynkov commented on src/main/java/com/example/telegrambotinsurance/service/HighLevelSendRequestService.java:
```diff
@@ -0,0 +1,135 @@
+	/**
+	 * class for generality of composing the http request string
+	 */
+	private enum TGApiRequest {
```
ну тут тоже по названию не оч понятно что за запрос, если по сути у тебя тут просто названия методов из апи телеги будут после изменений

klimenkoMA replied:
Окей, переделаю

spasynkov commented on src/main/java/com/example/telegrambotinsurance/modelbot/OutgoingMessage.java:
```diff
@@ -0,0 +1,22 @@
+@NoArgsConstructor
+@AllArgsConstructor
+
+public class OutgoingMessage {
```
вроде ж договаривались сделать просто `Message`?
а чтобы не путать с `Message` входящим - решили просто разложить эти два класса в разные пакеты :)
или нет?

klimenkoMA replied:
Я подумал, что с таким названием точно не будет путаницы. Нужно переделать?

spasynkov replied:
да не, это не критично)
просто помню вроде, что договорились по-одному сделать, а тут - другое :)

spasynkov commented on src/test/java/com/example/telegrambotinsurance/service/HighLevelSendRequestServiceTest.java:
```diff
@@ -0,0 +1,112 @@
+import java.util.Map;
+
+@SpringBootTest
+public class HighLevelSendRequestServiceTest {
```
если поменяешь название класса того - не забудь и тут поменять тоже

klimenkoMA replied:
Окей, поменяю

klimenkoMA replied:
Зарефакторил название класса

spasynkov commented on src/test/java/com/example/telegrambotinsurance/service/HighLevelSendRequestServiceTest.java:
```diff
@@ -0,0 +1,112 @@
+	 */
+	@Autowired
+	HighLevelSendRequestService highLevelSendRequestService;
+	@Autowired
+	SendRequestsService sendRequestsService;
```
а зачем это тут? у того класса есть свои тесты, и там все проверяется что надо

spasynkov commented on src/test/java/com/example/telegrambotinsurance/service/HighLevelSendRequestServiceTest.java:
```diff
@@ -0,0 +1,112 @@
+	HighLevelSendRequestService highLevelSendRequestService;
+	@Autowired
+	SendRequestsService sendRequestsService;
+	@Value("${bot.insurance.token}")
+	String token;
+	@Autowired
+	@Qualifier("BotsWithTokens")
+	Map<String, AbstractBot> bots;
```
поскольку токен в твоих методах не нужен (его можно достать из бота, которого в любом случае надо передавать) - то это все можно вынести в конструктор, а переменную оставить только с абстрактным ботом тут

klimenkoMA replied:
Пока не могу понять, как вынести в конструктор, но попробую

spasynkov replied:
да просто :)
```
private AbstractBot bot;

public HighLevelSendRequestServiceTest(@Value("${bot.insurance.token}")	String token, BotService service) {
    this.bot = service.findBotByToken(token);
}
```
у нас функционал по доставанию бота из мапы - уже реализован в соответствующем сервисе. поэтому нет смысла его дублировать тут, если можно воспользоваться уже готовым

klimenkoMA replied:
Спасибо, переделаю

klimenkoMA replied:
Добавил конструктор, переделал логику тестов

spasynkov commented on src/test/java/com/example/telegrambotinsurance/service/HighLevelSendRequestServiceTest.java:
```diff
@@ -0,0 +1,112 @@
+	/**
+	 * testing the "getChat" request with long chat ID
+	 */
+	@Test
+	public void integerMethodGetChatShouldReturnedJSONAnswer() {
+
+		long chatId = 871614990;
+
+		AbstractBot bot = bots.get(token);
+
+		JSONObject response = highLevelSendRequestService.getChat(bot, token, chatId);
+
+		// response must be not null
+		Assertions.assertNotNull(response);
+
+		// response must have "ok" status
+		Assertions.assertEquals("true", response.optString("ok"),
+				"Wrong response from TG API, field: ok = false ");
+	}
```
поскольку метода того с лонговым айдишником чата не надо нам - то и тест этот не нужен

klimenkoMA replied:
Окей, уберу

spasynkov commented on src/test/java/com/example/telegrambotinsurance/service/HighLevelSendRequestServiceTest.java:
```diff
@@ -0,0 +1,112 @@
+	@Test
+	public void stringMethodGetChatShouldReturnedJSONAnswer() {
+
+		String chatId = "871614990";
```
можно вынести в константу, если в нескольких методах использоваться будет тут

klimenkoMA replied:
Хорошо, сделаю

spasynkov commented on src/test/java/com/example/telegrambotinsurance/service/HighLevelSendRequestServiceTest.java:
```diff
@@ -0,0 +1,112 @@
+
+		String chatId = "871614990";
+
+		AbstractBot bot = bots.get(token);
```
как я уже написал, вынести получение бота в конструктор

klimenkoMA replied:
Принял, вынесу

spasynkov commented on src/test/java/com/example/telegrambotinsurance/service/HighLevelSendRequestServiceTest.java:
```diff
@@ -0,0 +1,112 @@
+	 */
+	@Test
+	public void sendMessageWithOutKeyBoardShouldReturnedJSONAnswer() {
+		String chatId = "1029944281";
```
почему тут айдишник чата другой? у нас же один чат для тестов сейчас

klimenkoMA replied:
это мой айдишник, в групповой чат не давал отправлять сообщения,  я в Постмане пробовал. Требуется права администратора для бота и в конфиге еще что-то через BotFather нужно поменять

spasynkov replied:
как это не давал отправлять, если у нас в чатике куча сообщений была от бота этого?
может тот второй чатик как-то поднастроить? чтоб самого бота не редактировать.
давай об этом в телеге поговорим, там Григорий создавал тот чатик

klimenkoMA replied:
Разобрался благодаря твоей наводке, исправлю айдишник :)

spasynkov commented on src/test/java/com/example/telegrambotinsurance/service/HighLevelSendRequestServiceTest.java:
```diff
@@ -0,0 +1,112 @@
+	 * testing the "sendMessage" request without keyboard
+	 */
+	@Test
+	public void sendMessageWithOutKeyBoardShouldReturnedJSONAnswer() {
```
`without` - это одно слово :)
как и `keyboard`
ну и по фен-шую, `JSONAnswer` надо бы заменить на `JsonAnswer`, чтоб по код-стайлу

klimenkoMA replied:
Принял, переделаю

spasynkov commented on src/test/java/com/example/telegrambotinsurance/service/HighLevelSendRequestServiceTest.java:
```diff
@@ -0,0 +1,112 @@
+	@Test
+	public void sendMessageWithOutKeyBoardShouldReturnedJSONAnswer() {
+		String chatId = "1029944281";
+		String text = "Don`t worry, be happy";
```
лучше все-таки писать что-то более ожидаемое, как для тестового сообщения :)
типа `test` или `test message`

klimenkoMA replied:
Хорошо, исправлю

spasynkov commented on src/test/java/com/example/telegrambotinsurance/service/HighLevelSendRequestServiceTest.java:
```diff
@@ -0,0 +1,112 @@
+		String chatId = "1029944281";
+		String text = "Don`t worry, be happy";
+		AbstractBot bot = bots.get(token);
+		OutgoingMessage message = new OutgoingMessage();
+		message.setStringChatId(chatId);
+		message.setText(text);
```
когда уберешь "лишний" чат айди - тут можно будет сделать одной строкой, в конструкторе передать

klimenkoMA replied:
Хорошо, сделаю

spasynkov reviewed (changes requested):
сори, но замечаний наверное больше, чем кода получилось :)

klimenkoMA committed:
#14 changes
- replaced a "TODO" in HighLevelSendRequestService class

klimenkoMA committed:
#14 changes
- renamed a HighLevelSendRequestService class to TelegramApi class

klimenkoMA committed:
#14 changes
- added a static modifier to LOGGER field in TelegramApi class

klimenkoMA committed:
#14 changes in TelegramApiTest class
- added a constructor to obtain instance of the bot
- deleted some fields
- rebuild all test methods
All tests are passed
